File: workflow/scripts/filter_krakenuniq.R
#This is a script for filtering KrakenUniq output. Run this script as:
#Rscipt filter_krakenuniq.R krakenuniq_output pathogens_list n_unique_kmers n_tax_reads

args = commandArgs(trailingOnly=TRUE)
krakenuniq_output<-as.character(args[1])
pathogens_list<-as.character(args[2])
n_unique_kmers<-as.integer(args[3])
n_tax_reads<-as.integer(args[4])


#Reading KrakenUniq output
df <- read.delim(krakenuniq_output,comment.char="#",header=TRUE)

#Filtering KrakenUniq output with respect to depth (taxReads) and breadth (kmers) of coverage
colnames(df)[1] <- "Pers_Reads"
df <- df[df$kmers>n_unique_kmers,]
df <- df[as.character(df$rank) == "species",]
df <- df[df$taxReads>n_tax_reads,]
write.table(df, file = paste0(krakenuniq_output,".filtered"), col.names=TRUE, row.names=FALSE, quote=FALSE, sep="\t")

#Overlapping the filtered KrakenUniq output with the list of pathogens
pathogens <- read.delim(pathogens_list, header=FALSE)
df <- df[as.character(df$taxID)%in%pathogens$V1,]
write.table(df, file = paste0(krakenuniq_output,".pathogens"), col.names=TRUE, row.names=FALSE, quote=FALSE, sep="\t")

File: workflow/scripts/filter_krakenuniq.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from sys import argv
import pandas as pd

krakenuniq_output = argv[1]

n_unique_kmers = int(argv[2])
n_tax_reads = int(argv[3])
organism_list = argv[4]

# Read and filter KrakenUniq output
kraken_output_df = pd.read_csv(krakenuniq_output, delimiter="\t", comment="#")
organism_list_df = pd.read_csv(organism_list, delimiter="\t", header=None)
print(organism_list_df)
print(f"Original data set dimensions: {kraken_output_df.shape}")

kraken_output_df = kraken_output_df[kraken_output_df["kmers"] > n_unique_kmers]

print(
    f"Data set dimensions after breadth of coverage filter: {kraken_output_df.shape}"
)
kraken_output_df = kraken_output_df[kraken_output_df["taxReads"] > n_tax_reads]
kraken_output_df = kraken_output_df[kraken_output_df["rank"] == "species"]
kraken_output_df = kraken_output_df.sort_values(by="%", ascending=False)

print(
    f"Data set dimensions after depth of coverage filter: {kraken_output_df.shape}"
)
kraken_output_df.to_csv(f"{krakenuniq_output}.filtered", sep="\t", index=False)

kraken_output_df = kraken_output_df[
    kraken_output_df.taxID.isin(organism_list_df.iloc[:, 0])
]
kraken_output_df.to_csv(
    f"{krakenuniq_output}.pathogens", sep="\t", index=False
)
